src/const-safety.md

```diff
@@ -0,0 +1,171 @@
+* Aggregates must not contain dangling `Scalar::Ptr` (e.g. in padding or unions) and additionally
+  follow the following aggregate-specific rules:
+    * Tuples, structs, arrays and slices are const-safe if all their fields are
+    const-safe.
+    * Enums are const-safe if they have a valid discriminant and the fields of the
+    active variant are const-safe.
+    * Unions are always const-safe; the data does not matter.
+
+
+For example:
+
+```rust
+static S: i32 = 0;
+const BAD: usize = unsafe { &S as *const i32 as usize };
+```
+
+Here, `S` is const-safe because `0` is a `Scalar::Bits`.  However, `BAD` is *not* const-safe because it is a `Scalar::Ptr`.
+Also in
+
+```rust
+static X: i32 = 42;
+static GOOD: *const i32 = &X;
+static BAD: *const i32 = { let x = 42; &x as *const i32 };
+```
+
+`BAD` is not const-safe, because it points to the local variable `x` which will be deallocated, while
+`GOOD` is const-safe, because it points to some memory that lives as long as the static itself.
+
+
+## Const safety check on code
+
+The purpose of the const safety check on code is to prohibit construction of
+non-const-safe values in safe code.  We can allow *almost* all runtime-safe operations,
+except for unconst operations -- which are all related to raw pointers:
+
+* Comparing raw pointers for (in)equality or order,
+* converting raw pointers to integers,
+* inspecting the bits of raw pointers (e.g. by hashing them).
+
+
+These operations are unconst because a `const fn` that
+(when called with the same arguments at runtime and compile-time)
+
+  * fails to run at compile time when it works perfectly fine at runtime
+  * produces a different result at compile time than at runtime
+
+is undefined behavior.
+
+`unsafe` blocks permit performing possibly unconst operations.
+At this point, it becomes the responsibility of the
+programmer to preserve const safety.  In particular, a *safe* `const fn` must
+always execute const-safely when called with const-safe arguments, and produce a
+const-safe result.  For example, the following function is const-safe (after
+some extensions of the miri engine that are already implemented in miri) even
+though it uses raw pointer operations:
+
+```rust
+const fn slice_eq(x: &[u32], y: &[u32]) -> bool {
+    if x.len() != y.len() {
+        return false;
+    }
+    // equal length and address -> memory must be equal, too
+    if unconst { x as *const [u32] as *const u32 == y as *const [u32] as *const u32 } {
+        return true;
+    }
+    // assume the following is legal const code for the purpose of this function
+    x.iter().eq(y.iter())
+}
+```
+
+On the other hand, the following function is *not* const-safe and hence it is considered a bug to mark it as such:
+
+```
+const fn ptr_eq<T>(x: &T, y: &T) -> bool {
+    unconst { x as *const T == y as *const T }
+}
+```
+
+If the function were invoked as `ptr_eq(&42, &42)` the result depends on the potential
+deduplication of the memory of the `42`s.
```

Comment:
Or rather, the result is that we get an error of the first kind: an unsupported operation.